Sources/SourceKitLSP/Swift/SwiftLanguageServer.swift

```diff
@@ -161,71 +161,100 @@ public final class SwiftLanguageServer: ToolchainLanguageServer {
     }
   }
 
-  /// Update lexical and syntactic tokens for the given `snapshot`.
-  /// Should only be called on `queue`.
+  /// Updates the lexical and syntactic tokens for the given `snapshot`.
+  /// Must be called on `self.queue`.
   private func updateLexicalAndSyntacticTokens(
     response: SKDResponseDictionary,
     for snapshot: DocumentSnapshot
   ) {
     dispatchPrecondition(condition: .onQueue(queue))
 
     let uri = snapshot.document.uri
+    let docTokens = updatedLexicalAndSyntacticTokens(response: response, for: snapshot)
+
+    do {
+      try documentManager.updateTokens(uri, tokens: docTokens)
+    } catch {
+      log("Updating lexical and syntactic tokens failed: \(error)", level: .warning)
+    }
+  }
+
+  /// Returns the updated lexical and syntactic tokens for the given `snapshot`.
+  private func updatedLexicalAndSyntacticTokens(
+    response: SKDResponseDictionary,
+    for snapshot: DocumentSnapshot
+  ) -> DocumentTokens {
+    var docTokens = snapshot.tokens
+
+    guard let offset: Int = response[keys.offset],
+          let length: Int = response[keys.length],
+          let start: Position = snapshot.positionOf(utf8Offset: offset),
+          let end: Position = snapshot.positionOf(utf8Offset: offset + length) else {
+      log("updateLexicalAndSyntacticTokens failed, no range found", level: .error)
+      return docTokens
+    }
+
+    let range = start..<end
+
+    // If the range is empty we don't have to (and shouldn't) update anything.
+    // This is important, since the substructure may be empty, causing us to
+    // unnecessarily remove all syntactic tokens.
+    guard !range.isEmpty else {
```

Comment:
Do you have a test case that requires this behavior? If yes, I would be interested what it is, if not, I think it would be good to add one. 😉

Reply:
Yes, `testEmptyEdit` requires this, there we have an empty range that returns an empty substructure too (apparently).

Reply:
Looking at it locally, an empty edit doesn’t have an `offset` and `length` in the response, so this should be handled by the `guard` above (which, in that case, shouldn’t log an error, I think 😉).

<details>
<summary>Empty edit through sourcekitd-test</summary>

```
./sourcekitd-test -req=open /Users/alex/Desktop/test.swift -- /Users/alex/Desktop/test.swift == -req=edit -offset=0 -length=0 -replace="" /Users/alex/Desktop/test.swift -print-raw-response -- /Users/alex/Desktop/test.swift
{
  key.request: source.request.editor.open,
  key.name: "/Users/alex/Desktop/test.swift",
  key.compilerargs: [
    "/Users/alex/Desktop/test.swift"
  ],
  key.sourcefile: "/Users/alex/Desktop/test.swift"
}
{
  key.request: source.request.editor.replacetext,
  key.name: "/Users/alex/Desktop/test.swift",
  key.compilerargs: [
    "/Users/alex/Desktop/test.swift"
  ],
  key.offset: 0,
  key.length: 0,
  key.sourcefile: "/Users/alex/Desktop/test.swift",
  key.sourcetext: ""
}
{
  key.diagnostic_stage: source.diagnostic.stage.swift.sema,
  key.syntaxmap: [
  ],
  key.substructure: [
  ],
  key.diagnostics: [
    {
      key.line: 2,
      key.column: 7,
      key.filepath: "/Users/alex/Desktop/test.swift",
      key.severity: source.diagnostic.severity.warning,
      key.id: "pbd_never_used",
      key.description: "initialization of immutable value 'x' was never used; consider replacing with assignment to '_' or removing it",
      key.diagnostic_stage: source.diagnostic.stage.swift.sema,
      key.fixits: [
        {
          key.offset: 15,
          key.length: 5,
          key.sourcetext: "_"
        }
      ],
      key.categories: [
        source.diagnostic.category.no_usage
      ]
    }
  ]
}
```

</details>

Reply:
The issue might be that `SKDResponseDictionary`'s subscript implementation

```swift
public subscript(key: sourcekitd_uid_t?) -> Int? {
  return Int(sourcekitd.api.variant_dictionary_get_int64(dict, key))
}
```

never returns `nil`, even if the values aren't set.

Reply:
@benlangmuir Should we add this check to the other subscripts too?

Reply:
Bool, Int and SKDResponseArray should be updated to use that approach, yes.  String and UID subscripts should already correct, because they already return NULL directly from the sourcekitd API.